<launch>
  <arg name="model" />

  <param name="use_sim_time" value="true"/>
  <param name="robot_description" textfile="$(find autolabor_description)/urdf/autolabor_mini.urdf" />
 
  <node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher" >
  </node>
  <node name="state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" >
  </node>

  <!-- robot_localization EKF node for the odom frame -->
  <node pkg="robot_localization" type="ekf_localization_node" name="robot_localization_ekf_node_odom"  clear_params="true">
    <param name="frequency" value="50" />
    <param name="sensor_timeout" value="0.2" />
    <param name="predict_to_current_time" value="true"/>
    <param name="two_d_mode" value="false" />
    <param name="publish_tf" value="true" />

    <param name="map_frame" value="map" />
    <param name="odom_frame" value="odom" />
    <param name="base_link_frame" value="base_link" />
    <param name="world_frame" value="odom" />

    <param name="print_diagnostics" value="true" />
    <param name="debug" value="false" />
    <param name="debug_out_file" value="debug_odom_ekf.txt" />

    <!-- The camera is bad at position, only use velocity here. -->
    <param name="odom0" value="camera_odom" />
    <rosparam param="odom0_config">[
      false, false, false, false, false, false,
      true, true, true, true, true, true,
      false, false, false]</rosparam>

    <!-- lidar gives a position information from tf. -->
    <param name="odom1" value="lidar_odom" />
    <rosparam param="odom1_config">[
      true, true, true, true, true, true,
      false, false, false, false, false, false,
      false, false, false]</rosparam>

    <!-- IMU handles the linear acceleration and angular velocity. -->
    <param name="imu0" value="imu_data"/>
    <param name="imu0_differential" value="false" />
    <param name="imu0_relative" value="false" />
    <param name="imu0_queue_size" value="10" />
    <param name="imu0_remove_gravitational_acceleration" value="true" />
    <rosparam param="imu0_config">[
      false, false, false, false, false, false,
      false, false, false, true, true, true,
      false, false, false]</rosparam>
  </node>
  
  <node name="slambox_odom_filter" pkg="slambox" type="slambox_odom_filter" output="screen" >
    <param name="vo_need_cov" value="true"/>
    <param name="vo_pos_cov" value="10" />
    <param name="vo_vel_cov" value="1" />
    <param name="lidar_need_cov" value="true"/>
    <param name="lidar_pos_cov" value="0.1" />
    <param name="lidar_vel_cov" value="10" />
  </node>

  <!-- Change frame_id in imu data -->
  <node pkg="slambox" type="IMU_filter" name="IMU_filter" output="log" >
    <param name="ori_cov" value="1"/>
  </node>

  <!-- <node pkg="tf" type="static_transform_publisher" name="robot_to_camera" args="0.0 0.0 0.0 0.0 0.0 0.0 /camera_pose_frame /base_link 200" /> -->
   <node pkg="tf" type="static_transform_publisher" name="sensor_init_to_odom"
         args="0.0 0.0 0.0 0.0 0.0 0.0 /sensor_init /odom 200" />
<!--  <node pkg="tf" type="static_transform_publisher" name="base_link_to_velodyne" args="0.0 0.0 0.0 0.0 0.0 0.0 /base_link /velodyne 200" />-->

  <node name="odometry_for_elev_map" pkg="topic_tools" type="transform"
        args="/odometry/filtered /odometry_for_elevation_map
        geometry_msgs/PoseWithCovarianceStamped
        'geometry_msgs.msg.PoseWithCovarianceStamped(header=m.header,pose=m.pose)' --import geometry_msgs" />

  <node pkg="elevation_mapping" type="elevation_mapping"
        name="elevation_mapping" output="screen">
    <rosparam command="load" file="$(find slambox)/config/slambox.yaml" />
    <rosparam command="load" file="$(find slambox)/config/map.yaml" />
    <!--    <rosparam command="load" file="$(find elevation_mapping)/config/sensor_processors/velodyne_HDL-32E.yaml" />-->
    <rosparam command="load" file="$(find elevation_mapping)/config/sensor_processors/perfect.yaml" />
  </node>

  <node name="rviz" pkg="rviz" type="rviz" args="-d $(find slambox)/rviz/mapping.rviz" />
  <node pkg="rosbag" type="play" name="player" output="screen"
        args="/home/gzy/bags/slambox/test.bag --clock -s 0 -r 1 --pause"/>
</launch>
